tools/logstash/lib/main.sc

```diff
@@ -0,0 +1,188 @@
+import akka.stream.IOResult
+import akka.util.ByteString
+import akka.stream.scaladsl._
+import ammonite.ops._
+import scala.concurrent.duration._
+import scala.concurrent.{Await, Future}
+import scala.util.{Try, Success, Failure}
+
+@tailrec def await[T](f: Future[T]): T = f.value match {
+  case None =>
+    Thread.sleep(10)
+    await(f)
+  case Some(v) =>
+    v.get
+}
+
+def renderTemplate(template: Path, vars: (String, String)*): String =
+  vars.foldLeft(read!(template)) { case (str, (env, value)) =>
+    str.replace(s"%${env}%", value)
+  }
+
+def gzipSource(input: Path, maxChunkSize: Int = 1024): Source[ByteString, Future[IOResult]] = {
+  FileIO.fromPath(input.toNIO).via(Compression.gunzip(maxChunkSize))
+}
+
+def bundleLogGzipped(masterPath: Path) =
+  masterPath / "dcos-marathon.service.gz"
+
+def bundleLogGunzipped(masterPath: Path) =
+  masterPath / "dcos-marathon.service"
+
+def warningLineSplitter(file: Path, warnLength: Int) =
+  Framing.delimiter(ByteString("\n"), 128000000, true).map { line =>
+    if (line.length >= warnLength)
+      println(s"WARNING!!! ${file} has a line length of ${line.length}")
+    line
+  }
+
+def unzipAndStripLogs(masters: Map[String, Path])(implicit mat: Materializer): Map[String, Path] = {
+  masters.map { case (master, masterPath) =>
+    val gunzippedFilePath = bundleLogGunzipped(masterPath)
+    if (!gunzippedFilePath.toIO.exists) {
+
+      val gzippedFilePath = bundleLogGzipped(masterPath)
+      println(s"Extracting ${gzippedFilePath} to ${gunzippedFilePath}")
+      val result = await {
+        gzipSource(gzippedFilePath)
+          .via(warningLineSplitter(gzippedFilePath, 128000))
+          .map { bytes => ByteString(util.stripAnsi(bytes.utf8String)) }
+          .intersperse(ByteString("\n"))
+          .runWith(FileIO.toPath(gunzippedFilePath.toNIO))
+      }
+      if (!result.wasSuccessful) {
+        println(s"WARNING! Error extracting ${gzippedFilePath}; ${result.status}. ${result.count} bytes were written.")
+      }
+    }
+    master -> gunzippedFilePath
+  }
+}
+
+def detectLogFormat(logFiles: Seq[Path])(implicit mat: Materializer): LogFormat = {
+  logFiles
+    .filter(_.toIO.exists)
+    .toSeq
+    .sortBy(_.toIO.length)
+    .reverse
+    .toStream
+    .map { input =>
+      val result = Try {
+        val linesSample = await(
+          FileIO.fromPath(input.toNIO)
+            .via(warningLineSplitter(input, 128000))
+            .take(100)
+            .map(_.utf8String)
+            .runWith(Sink.seq))
+
+        val maybeCodec = (for {
+          line <- linesSample.take(100)
+          codec <- LogFormat.all if codec.matches(line)
+        } yield codec).headOption
+
+        maybeCodec match {
+          case Some(codec) => codec
+          case _ =>
+            println(s"Couldn't find a codec for these lines:")
+            println()
+            linesSample.foreach(println)
+            // sys.exit(1)
+            ???
```

Comment:
What is this about?

Reply:
`???` throws not implemented exception. Because codec not implemented.

Reply:
Hm, ok. I'd be more specific but let's merge this 🙂.